src/Canvas.cc

```diff
@@ -51,16 +52,17 @@ Canvas::Initialize(Nan::ADDON_REGISTER_FUNCTION_ARGS_TYPE target) {
 
   // Prototype
   Local<ObjectTemplate> proto = ctor->PrototypeTemplate();
+  proto->SetInternalFieldCount(1);
```

Comment:
Hey @zbjornson, thanks for linking to this PR.

While this is just fine as a Band-Aid for the problem, I'd like to note several potential problems that would stem from this implementation. First, philosophically, the prototype object should not be an instance of the class -- that is, it shouldn't have internal slots like instances do. JavaScript got this wrong with `Array.prototype` and `Function.prototype` (`Array.isArray(Array.prototype)` and `typeof Function.prototype === 'function'` both return true), but they have rectified this problem with future additions to the language with a note in the [spec](https://tc39.github.io/ecma262/#sec-properties-of-the-array-prototype-object):

> The Array prototype object is specified to be an Array exotic object to ensure compatibility with ECMAScript code that was created prior to the ECMAScript 2015 specification.

More concretely, although accessing `Canvas.prototype.type` will not throw, other code fragments e.g. `Object.create(Canvas.prototype).type` will still throw, I believe. Even worse (but more unlikely), `Object.setPrototypeOf(anotherObjectWithInternalFields, Canvas.prototype).type` may cause memory issues, as the type of the object pointer wrapped in V8 object is untyped.

If a Band-Aid is all that is sought here, then please disregard my suggestions below to make this more robust. Either suggestion will unfortunately incur quite some work, so it's totally up to you regarding what's best for this project.

1. Write a pure JS wrapper around the JS object created in C++, and implement the property getters/setters as ordinary functions (e.g. `getLength()`, `setLength()`). This means that you would do all necessary type checking in JS, so in C++ you can feel free to assert() on predicates. The C++ wrapper would also be hidden in a symbol, so we can tell any user that get curious and call the prototype methods w/o context, well, don't try to access things you are not supposed to like hidden properties. This is generally what we do in Node.js.

2. Check if `info.Holder()` is an instance of the `FunctionTemplate`, and throw an error if it is not. This is what Chrome [does](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/platform/bindings/V8PerIsolateData.cpp?l=310&rcl=4f7fcf34302b3c09572d978bcd3e58fe6eb5c703) for DOM, but it would require holding a reference to the FunctionTemplate, always. If you were to go down this route, I would put the Persistent reference to the FunctionTemplate in a `std::unordered_map` keyed on `Isolate*`.

A lot of this is based on speculation, so please let me know if you found a better solution to all of this!

Reply:
Thanks so much for the detailed and helpful info!

>  the prototype object should not be an instance of the class -- that is, it shouldn't have internal slots like instances do

Oops, `proto->SetInternalFieldCount(1)` was leftover from before I added the signatures to the `SetAccessor`s, and no value is set for the slot anyway. Will remove in a moment.

As far as the proto being an instance of the class: my understanding is that internal slots are usable for anything and have no effect on the behavior of the object (they're just C++-accessible storage). `Canvas.prototype instanceof Canvas === false`, so I think we're okay?

> `Object.create(Canvas.prototype).type` will still throw, I believe. ... `Object.setPrototypeOf(anotherObjectWithInternalFields, Canvas.prototype).type` may cause memory issues

Both of these throw the "incompatible receiver" error (as desired; same as `Canvas.prototype.type`), and neither have the assertion failure.

> Check if info.Holder() is an instance of the FunctionTemplate

I think this is what v8 does actually, when `SetAccessor` is used with an `AccessorSignature`:

> *The signature describes valid receivers for the accessor and is used to perform implicit instance checks against them. If the receiver is incompatible (i.e. is not an instance of the constructor as defined by FunctionTemplate::HasInstance()), an implicit TypeError is thrown and no callback is invoked.*

Am I reading that right? If so, then I think this PR is actually a sturdy route forward.

Reply:
I wasn't familiar with how AccessorSignature worked... and it looks like that's it! This looks good to me.

Reply:
Thank you very much for reviewing!